File: python/algorithms/searching/ex1_linear_search.py
"""
Exercise: Linear Search (Easy)

Implement linear search to find the index of a target element in a list.
Linear search checks each element sequentially from left to right until the
target is found or the end of the list is reached.

Example:
Input: arr = [4, 2, 7, 1, 9, 5], target = 7
Output: 2

Example (not found):
Input: arr = [4, 2, 7, 1, 9, 5], target = 10
Output: -1

Time Complexity: O(n) - must potentially check every element
Space Complexity: O(1) - only uses a constant amount of extra space
"""

from typing import List


def linear_search(arr: List[int], target: int) -> int:
    """
    Searches for a target element in a list using linear search.

    Args:
        arr: The list to search
        target: The element to find

    Returns:
        The index of the target if found, -1 otherwise
    """
    # TODO: Implement linear search
    # Hint 1: Use enumerate to iterate through elements with their indices
    # Hint 2: Compare each element with the target
    # Hint 3: Return the index when found
    # Hint 4: Return -1 if the loop completes without finding the target
    raise NotImplementedError("Not implemented yet")


if __name__ == "__main__":
    # Example usage
    test_arr = [4, 2, 7, 1, 9, 5]
    target = 7
    result = linear_search(test_arr, target)
    print(f"Index of {target} in {test_arr}: {result}")

File: rust/algorithms/src/searching/ex1_linear_search.rs
/// Exercise 1: Linear Search
/// Difficulty: Easy
///
/// Implement linear search to find the index of a target element in an array.
/// Linear search checks each element sequentially until the target is found or
/// the array is exhausted.
///
/// # Examples
///
/// ```
/// use algorithms::searching::ex1_linear_search::linear_search;
///
/// let arr = vec![5, 2, 8, 1, 9];
/// assert_eq!(linear_search(&arr, 8), Some(2));
/// assert_eq!(linear_search(&arr, 7), None);
/// ```
///
/// Time Complexity: O(n) - must potentially check every element
/// Space Complexity: O(1) - only uses a constant amount of extra space

/// Performs linear search on the array to find the target element.
///
/// # Arguments
///
/// * `arr` - The array to search
/// * `target` - The element to find
///
/// # Returns
///
/// The index of the target element, or None if not found
pub fn linear_search(arr: &[i32], target: i32) -> Option<usize> {
    // TODO: Implement linear search
    // Hint: Iterate through the array and check each element
    unimplemented!("Not implemented yet")
}

/// Alternative signature that searches within a specific range.
///
/// # Arguments
///
/// * `arr` - The array to search
/// * `target` - The element to find
/// * `start` - The starting index (inclusive)
/// * `end` - The ending index (exclusive)
///
/// # Returns
///
/// The index of the target element, or None if not found
pub fn linear_search_range(arr: &[i32], target: i32, start: usize, end: usize) -> Option<usize> {
    // TODO: Implement ranged linear search
    unimplemented!("Not implemented yet")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_element_found() {
        let arr = vec![5, 2, 8, 1, 9];
        assert_eq!(linear_search(&arr, 8), Some(2));
    }

    #[test]
    fn test_element_not_found() {
        let arr = vec![5, 2, 8, 1, 9];
        assert_eq!(linear_search(&arr, 7), None);
    }

    #[test]
    fn test_first_element() {
        let arr = vec![10, 20, 30, 40];
        assert_eq!(linear_search(&arr, 10), Some(0));
    }

    #[test]
    fn test_last_element() {
        let arr = vec![10, 20, 30, 40];
        assert_eq!(linear_search(&arr, 40), Some(3));
    }

    #[test]
    fn test_single_element() {
        let arr = vec![42];
        assert_eq!(linear_search(&arr, 42), Some(0));
    }

    #[test]
    fn test_single_element_not_found() {
        let arr = vec![42];
        assert_eq!(linear_search(&arr, 1), None);
    }

    #[test]
    fn test_empty_array() {
        let arr: Vec<i32> = vec![];
        assert_eq!(linear_search(&arr, 5), None);
    }

    #[test]
    fn test_duplicate_elements() {
        let arr = vec![3, 5, 3, 7, 3];
        // Should return the first occurrence
        assert_eq!(linear_search(&arr, 3), Some(0));
    }

    #[test]
    fn test_ranged_search() {
        let arr = vec![1, 2, 3, 4, 5, 6, 7];
        // Search for 5 between index 2 and 6
        assert_eq!(linear_search_range(&arr, 5, 2, 6), Some(4));
    }

    #[test]
    fn test_ranged_search_not_in_range() {
        let arr = vec![1, 2, 3, 4, 5, 6, 7];
        // Search for 7 between index 0 and 5 (7 is at index 6)
        assert_eq!(linear_search_range(&arr, 7, 0, 5), None);
    }
}
